rfcs/45-stageless.md

```diff
@@ -0,0 +1,980 @@
+    timer.tick(time.delta());
+}
+
+// This function can serve as a condition because it does not have mutable access and it returns `bool`.
+fn construction_timer_not_finished(timer: Res<ConstructionTimer>) -> bool {
+    !timer.finished()
+}
+
+// As long as they're immutable, you can use queries, event readers and resources with arbitrarily complex internal logic.
+fn too_many_enemies(
+    population_query: Query<(), With<Enemy>>,
+    population_limit: Res<PopulationLimit>,
+) -> bool {
+    population_query.iter().count() > population_limit
+}
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        // You can add functions with read-only system parameters as conditions.
+        .add_set(GameSet::Construction.run_if(construction_timer_not_finished))
+        .add_systems(
+            systems![tick_construction_timer, update_construction_progress]
+                .in_set(GameSet::Construction))
+        .add_system(mitigate_meltdown.run_if(too_many_enemies))
+        // You can use closures for simple one-off conditions.
+        .add_system(spawn_more_enemies.run_if(|difficulty: Res<Difficulty>| difficulty >= 9000))
+        // `resource_exists` and `resource_equals` are helper functions that produce
+        // new closures that can be used as conditions.
+        .add_system(gravity.run_if(resource_exists(Gravity)))
+        // The systems in this set won't run if `Paused` is `true`.
+        .add_set(GameSet::Physics.run_if(resource_equals(Paused(false))))
+        .run();
+}
+```
+
+### Exclusive systems
+
+As much as everyone loves to see systems running in parallel, sometimes a little `&mut World` is necessary.
+Previous versions of Bevy treated these "exclusive" systems as a separate concept, with their own set of types and traits.
+They couldn't have `Local` params and could only be inserted at specific points of a stage.
+
+That is no longer the case.
+Exclusive systems are now just regular systems.
+They can have `Local` params and be scheduled wherever you want.
+
+Command application is now signaled through an exclusive system called `apply_system_buffers`.
+You can add instances of this system anywhere in your schedule.
+If one system depends on the effects of commands from another, make sure an `apply_system_buffers` appears somewhere between them.
+
+```rust
+use bevy::prelude::*;
+
+#[derive(SystemLabel)]
+enum MySystems {
+    /* ... */
+    FlushProjectiles,
+    /* ... */
+}
+
+struct ProjectilePlugin;
+
+impl Plugin for ProjectilePlugin {
+    fn build(app: &mut App) {
+        app
+        .add_system(spawn_projectiles.before(CoreSystems::FlushPostUpdate))
+        .add_systems(
+            chain![
+                check_if_projectiles_hit,
+                despawn_projectiles_that_hit,
+                // wherever you want commands to be applied, insert an instance of `apply_system_buffers`
+                apply_system_buffers.named(MySystems::FlushProjectiles)

```

Comment:
What does `.named` do and why is it necessary here?

Reply:
It gives the instance of the system a unique name.

You can have multiple instances of the same underlying function, but we need a way to tell them apart.